File: SEP3/SEP3 Project/DataAccessTier/src/main/java/dk/via/nbnp/databaseserver/domain/Conversation.java
package dk.via.nbnp.databaseserver.domain;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table
public class Conversation {

    @Id
    @SequenceGenerator(
            name="conversation_id_sequence",
            sequenceName = "conversation_id_sequence",
            allocationSize = 1
    )
    @GeneratedValue(
            strategy = GenerationType.SEQUENCE,
            generator = "conversation_id_sequence"
    )
    private Long id;

    @JsonFormat(pattern = "yyyy-MM-dd hh:mm:ss")
    private LocalDateTime dateTimeOfStart;

    @ManyToOne
    @JoinColumn(name = "buyerId")
    private User buyer;

    @ManyToOne
    @JoinColumn(name = "sellerId")
    private User seller;

    @ManyToOne
    @JoinColumn(name = "itemId")
    private Item item;

    @OneToMany(fetch = FetchType.EAGER)
    @JsonIgnore
    @JoinColumn(name = "conversationId")
    private List<Message> messageList;

    public Conversation(){}

    public Conversation(User buyer, User seller, Item item){
        this.buyer = buyer;
        this.seller = seller;
        this.item = item;
        this.dateTimeOfStart = LocalDateTime.now();
        messageList = new ArrayList<>();
    }

    public Item getItem() {
        return item;
    }
    public User getBuyer() {
        return buyer;
    }
    public User getSeller() {
        return seller;
    }
    public Long getId() {
        return id;
    }
    public LocalDateTime getDateTimeOfStart() {
        return dateTimeOfStart;
    }
    public List<Message> getMessageList() {
        return messageList;
    }
}

File: SEP3/SEP3 Project/BusinessLogicTier/Domain/Models/Conversation.cs
namespace Domain.Models;

public class Conversation
{
    public long Id { get; init;}
    public User Buyer { get; init;}
    public User Seller { get; init;}
    public Item Item { get; init;}
    public List<Message> MessageList { get; init;}
    public DateTime DateTimeOfStart { get; init;}

    // public Conversation(long id, User buyer, User seller, Item item, List<Message> messageList, DateTime dateTimeOfStart)
    // {
    //     Id = id;
    //     Buyer = buyer;
    //     Seller = seller;
    //     Item = item;
    //     MessageList = messageList;
    //     DateTimeOfStart = dateTimeOfStart;
    // }

}